File: app/src/main/java/com/mainstreetcode/teammate/viewmodel/gofers/EventGofer.java
package com.mainstreetcode.teammate.viewmodel.gofers;

import android.annotation.SuppressLint;
import android.arch.core.util.Function;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.util.DiffUtil;

import com.google.android.gms.location.places.Place;
import com.mainstreetcode.teammate.R;
import com.mainstreetcode.teammate.model.BlockedUser;
import com.mainstreetcode.teammate.model.Event;
import com.mainstreetcode.teammate.model.Guest;
import com.mainstreetcode.teammate.model.Identifiable;
import com.mainstreetcode.teammate.model.User;
import com.mainstreetcode.teammate.repository.GuestRepository;
import com.mainstreetcode.teammate.util.ErrorHandler;
import com.mainstreetcode.teammate.util.ModelUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Date;
import java.util.Iterator;
import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Single;
import io.reactivex.functions.Consumer;

public class EventGofer extends TeamHostingGofer<Event> {

    private boolean isSettingLocation;
    private final Function<Guest, Single<Guest>> rsvpFunction;
    private final Function<Event, Flowable<Event>> getFunction;
    private final Function<Event, Single<Event>> deleteFunction;
    private final Function<Event, Single<Event>> updateFunction;
    private final GuestRepository guestRepository;

    @SuppressLint("CheckResult")
    @SuppressWarnings("ResultOfMethodCallIgnored")
    public EventGofer(Event model,
                      Consumer<Throwable> onError,
                      Flowable<BlockedUser> blockedUserFlowable,
                      Function<Event, Flowable<Event>> getFunction,
                      Function<Event, Single<Event>> upsertFunction,
                      Function<Event, Single<Event>> deleteFunction,
                      Function<Guest, Single<Guest>> rsvpFunction) {
        super(model, onError);
        this.getFunction = getFunction;
        this.rsvpFunction = rsvpFunction;
        this.updateFunction = upsertFunction;
        this.deleteFunction = deleteFunction;
        this.guestRepository = GuestRepository.getInstance();

        items.addAll(model.asItems());
        items.add(model.getTeam());

        blockedUserFlowable.subscribe(this::onUserBlocked, ErrorHandler.EMPTY);
    }

    public void setSettingLocation(boolean settingLocation) {
        isSettingLocation = settingLocation;
    }

    public boolean isSettingLocation() {
        return isSettingLocation;
    }

    public String getToolbarTitle(Fragment fragment) {
        return model.isEmpty()
                ? fragment.getString(R.string.create_event)
                : fragment.getString(R.string.edit_event, model.getName());
    }

    @Nullable
    @Override
    public String getImageClickMessage(Fragment fragment) {
        if (hasPrivilegedRole()) return null;
        return fragment.getString(R.string.no_permission);
    }

    @Override
    Flowable<DiffUtil.DiffResult> fetch() {
        if (isSettingLocation) return Flowable.empty();
        Flowable<List<Identifiable>> eventFlowable = getFunction.apply(model).map(Event::asIdentifiables);
        Flowable<List<Identifiable>> guestsFlowable = guestRepository.modelsBefore(model, new Date()).map(ModelUtils::asIdentifiables);
        Flowable<List<Identifiable>> sourceFlowable = Flowable.concatDelayError(Arrays.asList(eventFlowable, guestsFlowable));
        return Identifiable.diff(sourceFlowable, this::getItems, this::preserveItems);
    }

    Single<DiffUtil.DiffResult> upsert() {
        Single<List<Identifiable>> source = updateFunction.apply(model).map(Event::asIdentifiables);
        return Identifiable.diff(source, this::getItems, this::preserveItems);
    }

    public Single<DiffUtil.DiffResult> rsvpEvent(boolean attending) {
        Single<List<Identifiable>> single = rsvpFunction.apply(Guest.forEvent(model, attending)).map(Collections::singletonList);
        return Identifiable.diff(single, this::getItems, (staleCopy, singletonGuestList) -> {
            staleCopy.removeAll(singletonGuestList);
            staleCopy.addAll(singletonGuestList);
            return staleCopy;
        });
    }

    public Single<Integer> getRSVPStatus() {
        return Flowable.fromIterable(items)
                .filter(identifiable -> identifiable instanceof Guest)
                .cast(Guest.class)
                .filter(Guest::isAttending)
                .map(Guest::getUser)
                .filter(getSignedInUser()::equals)
                .collect(ArrayList::new, List::add)
                .map(List::isEmpty)
                .map(notAttending -> notAttending ? R.drawable.ic_event_available_white_24dp : R.drawable.ic_event_busy_white_24dp);
    }

    Completable delete() {
        return deleteFunction.apply(model).toCompletable();
    }

    public Single<DiffUtil.DiffResult> setPlace(Place place) {
        isSettingLocation = true;
        model.setPlace(place);
        return Identifiable.diff(Single.just(model.asIdentifiables()), this::getItems, this::preserveItems).doFinally(() -> isSettingLocation = false);
    }

    private void onUserBlocked(BlockedUser blockedUser) {
        if (!blockedUser.getTeam().equals(model.getTeam())) return;

        Iterator<Identifiable> iterator = items.iterator();

        while (iterator.hasNext()) {
            Identifiable identifiable = iterator.next();
            if (!(identifiable instanceof Guest)) continue;
            User blocked = blockedUser.getUser();
            User guestUser = ((Guest) identifiable).getUser();

            if (blocked.equals((guestUser))) iterator.remove();
        }
    }
}

File: app/src/main/java/com/mainstreetcode/teammate/viewmodel/gofers/EventGofer.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.viewmodel.gofers

import android.annotation.SuppressLint
import android.location.Address
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DiffUtil
import com.mainstreetcode.teammate.R
import com.mainstreetcode.teammate.model.BlockedUser
import com.mainstreetcode.teammate.model.Event
import com.mainstreetcode.teammate.model.Guest
import com.mainstreetcode.teammate.repository.GuestRepo
import com.mainstreetcode.teammate.repository.RepoProvider
import com.mainstreetcode.teammate.util.ErrorHandler
import com.mainstreetcode.teammate.util.FunctionalDiff
import com.mainstreetcode.teammate.util.asDifferentiables
import com.tunjid.androidbootstrap.recyclerview.diff.Differentiable
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers.mainThread
import io.reactivex.schedulers.Schedulers
import java.util.*

class EventGofer @SuppressLint("CheckResult")
constructor(
        model: Event,
        onError: (Throwable) -> Unit,
        blockedUserFlowable: Flowable<BlockedUser>,
        private val getFunction: (Event) -> Flowable<Event>,
        private val updateFunction: (Event) -> Single<Event>,
        private val deleteFunction: (Event) -> Single<Event>,
        private val rsvpFunction: (Guest) -> Single<Guest>
) : TeamHostingGofer<Event>(model, onError) {

    var isSettingLocation: Boolean = false
    private val guestRepository: GuestRepo = RepoProvider.forRepo(GuestRepo::class.java)

    val rsvpStatus: Single<Int>
        get() = Single.fromCallable {
            ArrayList(items)
                    .filterIsInstance(Guest::class.java)
                    .filter(Guest::isAttending)
                    .map(Guest::user)
                    .filter { signedInUser == it }
                    .let {
                        if (it.isEmpty()) R.drawable.ic_event_available_white_24dp
                        else R.drawable.ic_event_busy_white_24dp
                    }
        }.subscribeOn(Schedulers.io()).observeOn(mainThread())

    init {

        items.addAll(model.asDifferentiables())
        items.add(model.team)

        blockedUserFlowable.subscribe(this::onUserBlocked, ErrorHandler.EMPTY::invoke)
    }

    fun getToolbarTitle(fragment: Fragment): CharSequence {
        return when {
            model.isEmpty -> fragment.getString(R.string.create_event)
            else -> model.name
        }
    }

    override fun getImageClickMessage(fragment: Fragment): String? =
            if (hasPrivilegedRole()) null else fragment.getString(R.string.no_permission)

    override fun fetch(): Flowable<DiffUtil.DiffResult> {
        if (isSettingLocation) return Flowable.empty()
        val eventFlowable = getFunction.invoke(model).map(Event::asDifferentiables)
        val guestsFlowable = guestRepository.modelsBefore(model, Date()).map(::asDifferentiables)
        val sourceFlowable = Flowable.concatDelayError(listOf(eventFlowable, guestsFlowable))
        return FunctionalDiff.of(sourceFlowable, items, this::preserveItems)
    }

    override fun upsert(): Single<DiffUtil.DiffResult> {
        val source = updateFunction.invoke(model).map(Event::asDifferentiables)
        return FunctionalDiff.of(source, items, this::preserveItems)
    }

    fun rsvpEvent(attending: Boolean): Single<DiffUtil.DiffResult> {
        val single: Single<List<Differentiable>> = rsvpFunction.invoke(Guest.forEvent(model, attending)).map { listOf(it) }

        return FunctionalDiff.of(single, items) { staleCopy, singletonGuestList ->
            (staleCopy - singletonGuestList) + singletonGuestList
        }
    }

    override fun delete(): Completable = deleteFunction.invoke(model).ignoreElement()

    fun setAddress(address: Address): Single<DiffUtil.DiffResult> {
        isSettingLocation = true
        model.setAddress(address)
        return FunctionalDiff.of(Single.just(model.asDifferentiables()), items, this::preserveItems)
                .doFinally { isSettingLocation = false }
    }

    private fun onUserBlocked(blockedUser: BlockedUser) {
        if (blockedUser.team != model.team) return

        val iterator = items.iterator()

        while (iterator.hasNext()) {
            val identifiable = iterator.next() as? Guest ?: continue
            val blocked = blockedUser.user
            val guestUser = identifiable.user

            if (blocked == guestUser) iterator.remove()
        }
    }
}
